lib/resty/websocket/client.lua

```diff
@@ -215,7 +232,11 @@ function _M.connect(self, uri, opts)
         return nil, "bad HTTP response status line: " .. header
     end
 
-    return 1
+    if not keep_response then
+        header = nil
+    end
+
+    return 1, nil, header
```

Comment:
We won't return the response headers to the caller when the socket was obtained from the keepalive pool as it stands. We either should add a test to confirm this is the intended behavior or fix said behavior and return the headers on reused sockets as well.

Reply:
Ah good point. Maybe in the case of connection reuse we should return a token value (could just be the string `reused connection`) instead of the response string so that the caller can be aware. What do you think?

Reply:
I worry that the returned `headers` value will be directly injected into some logic by users. But there is no way to distinguish this token from actual headers without knowing about it _and_ comparing it... It feels slightly contrived. Why not storing it? In the main thread that is the intent: "saving the upstream handshake response", why not doing that?

Reply:
So, I realize that I wasn't very clear about this. The stated "saving the upstream handshake response" goal was really just in reference to my initial implementation idea, which was to store the response on the client object (`self.response = headers` or similar). When I wrote that I actually hadn't considered the keepalive case.

For keepalive, I can't think of any feasible way to store the response without solving the problem of how to reference and manage the lifecycle of this string. Our options would be:

1. Storing it in the client object: the lifecycle of the `resty.websocket.client` object doesn't necessarily align with that of the (potentially-reused) keepalive connection, so if we persist the response there it won't be available when instantiating and connecting a new client (`client.new() + client:connect()`).

2. Storing it elsewhere: the only other option is to store the response in a module-level var (either private or on the `_M` module ns). We can't reference it by the client object (for the same reasons as option 1), so that means we have to derive a consistent keepalive pool name ourselves (if the user hasn't supplied one), which has all sorts of pitfalls. Then we have no straightforward way to GC it.

We could lengthen the return signature to make things slightly less ambiguous, returning `1, nil, nil, "reused connection"`. I don't see that as much of an improvement _personally_, but I'm not at all opposed to going that route if you think it's better from an API perspective.

Reply:
There are a couple other decision points to this feature that are more-or-less related:

> Why is this opt-in? Is `keep_response` flag even necessary at all? Can't we just always return it?

Answer is that as a newbie OpenResty contributor I was trying to be as conservative as possible with regards to backwards compat, but I _do_ think there's a case to be made that it's relatively safe to add a 3rd return value to `connect()` without making it opt-in, given that lua's flexibility and coding conventions make this a non-issue in most cases.

> If we're keeping it opt-in, `keep_response` is potentially misleading

Sure, I'd be happy to rename this to something else. `return_response` maybe?